tfx/orchestration/experimental/optimizations/beam_fusion.py

```diff
@@ -0,0 +1,197 @@
+
+class BeamFusionOptimizer(object):
+  """Optimizer for TFX pipelines, utilizing Beam Fusion"""
+
+  def __init__(self, pipeline: Pipeline):
+    self.pipeline = pipeline
+
+  def _topologically_sort(self, components: List[base_node.BaseNode],
+                          sources: List[base_node.BaseNode]):
+
+    sorted_components = []
+    current_layer = [c for c in components if c in sources]
+    visited = set(current_layer)
+
+    # Sorts component in topological order.
+    while current_layer:
+      next_layer = []
+      for component in sorted(current_layer, key=lambda c: c.id):
+        sorted_components.append(component)
+        for child in component.downstream_nodes:
+          if child in components and child not in visited:
+            next_layer.append(child)
+            visited.add(child)
+
+      current_layer = next_layer
+
+    return sorted_components
+
+  def _get_intersecting_subgraph(self,
+                                 intersecting_component: base_node.BaseNode,
+                                 fuseable_subgraphs: (
+                                     List[List[base_node.BaseNode]])):
+
+    intersecting_subgraph = None
+    for subgraph in fuseable_subgraphs:
+      if intersecting_component in subgraph:
+        intersecting_subgraph = subgraph
+        break
+
+    return intersecting_subgraph
+
+  def _is_beam_component(self, component: base_node.BaseNode):
+    return issubclass(component.EXECUTOR_SPEC.executor_class,
+                      base_executor.FuseableBeamExecutor)
+
+  def _is_fuseable(self, child: base_node.BaseNode,
+                   current_subgraph: List[base_node.BaseNode]):
+    is_fuseable = True
+
+    # Conduct a BFS to ensure none of the child's ancestors (other than its'
+    # immediate parents and those parents' ancestors) are in current_subgraph.
+    current_layer = []
+    for parent in child.upstream_nodes:
+      if not parent in current_subgraph:
+        current_layer.append(parent)
+
+    while(is_fuseable and current_layer):
+      next_layer = []
+      for component in current_layer:
+        for parent in component.upstream_nodes:
+          if parent in current_subgraph:
+            is_fuseable = False
+            break
+          next_layer.append(parent)
+
+      current_layer = next_layer
+
+    return is_fuseable
+
+  def get_fuseable_subgraphs(self):
+    """Returns a list of fuseable Beam component subgraphs in topological order.
+
+    Conducts multiple BFS searches to build out fuseable Apache Beam component
+    subgraphs. Each subgraph S must meet the following correctness constraints:
+    1. Each subgraph must be a directed acyclic graph
+    2. All the components in the subgraph must be Apache Beam Components
+    3. Consider a component in a subgraph being explored. None of the
+       component's ancestors (other than its' immediate parents and those
+       parents' ancestors) should be in the current subgraph being explored.
```

Comment:
I feel like the phrasing is a bit awkward because of the parentheses? My understanding is the condition can be rewritten as "if the component has parents that are not in the current subgraph, then the ancestors of those parents must not be in the current subgraph".

Reply:
Haha you're right, I replaced it with your suggested phrasing. Hopefully it is clearer now to a new reader.